src/emulator/dataconnect/pgliteServer.ts

```diff
@@ -130,18 +128,89 @@ export class PostgresServer {
     fs.writeFileSync(exportPath, new Uint8Array(arrayBuff));
   }
 
+  private async migrateDb(pgliteArgs: PGliteOptions): Promise<PGlite> {
+    if (!pgliteArgs.dataDir) {
+      throw new FirebaseError("Cannot migrate database without a data directory.");
+    }
+    const dataDir = pgliteArgs.dataDir;
+
+    // 1. Import old PGlite and pgDump
+    const { PGlite: PGlite02 } = await dynamicImport("pglite-2");
+    const pgDump = (await dynamicImport("@electric-sql/pglite-tools/pg_dump")).pgDump;
+
+    // 2. Open old DB with old PGlite
+
+    logger.info("Opening database with Postgres 16...");
+    const extensions = await this.getExtensions();
+    const oldDb = new PGlite02({ dataDir, extensions });
+    await oldDb.waitReady;
+
+    const oldVersion = await (oldDb as PGlite).query<{ version: string }>("SELECT version();");
+    logger.debug(`Old database version: ${oldVersion.rows[0].version}`);
+    if (!oldVersion.rows[0].version.includes("PostgreSQL 16")) {
+      await oldDb.close();
+      throw new FirebaseError("Migration started, but DB version is not PostgreSQL 16.");
+    }
+
+    // 3. Dump data
+    logger.info("Dumping data from old database...");
+    const dumpDir = await oldDb.dumpDataDir("none");
+    const tempOldDb = await PGlite02.create({
+      loadDataDir: dumpDir,
+      extensions,
+    });
+
+    const dumpResult = await pgDump({ pg: tempOldDb, args: ["--verbose", "--verbose"] });
+    await tempOldDb.close();
+    await oldDb.close();
+
+    // 4. Nuke old data directory
```

Comment:
Can we do this step last so that if something goes wrong during the import, the developer can still go back to an older CLI version? 

Even better, what if we create a subdirectory in dataDir, say `pg17` and only write to that subdirectory starting from this PR? In this way, the old and new CLI versions should not mess with each other. Ever time we change the on-disk format we can move to a different subdir name.

Reply:
Thanks, this looks great. I think we can make it even better by leaving existing data prefixless instead of `pg16`. In this way, downgrades are 100% safe out of the box without changing the dataDir settings.